pumpkin/src/command/commands/cmd_teleport.rs

```diff
@@ -0,0 +1,276 @@
+use async_trait::async_trait;
+use pumpkin_core::math::vector3::Vector3;
+use pumpkin_core::text::TextComponent;
+
+use crate::command::arg_entities::{parse_arg_entities, EntitiesArgumentConsumer};
+use crate::command::arg_entity::{parse_arg_entity, EntityArgumentConsumer};
+use crate::command::arg_position_3d::{parse_arg_position_3d, Position3DArgumentConsumer};
+use crate::command::arg_rotation::{parse_arg_rotation, RotationArgumentConsumer};
+use crate::command::tree::CommandTree;
+use crate::command::tree_builder::{argument, literal, require};
+use crate::command::InvalidTreeError;
+use crate::command::{tree::ConsumedArgs, CommandExecutor, CommandSender};
+
+const NAMES: [&str; 2] = ["teleport", "tp"];
+const DESCRIPTION: &str = "Teleports entities, including players."; // todo
+
+/// position
+const ARG_LOCATION: &str = "location";
+
+/// single entity
+const ARG_DESTINATION: &str = "destination";
+
+/// multiple entities
+const ARG_TARGETS: &str = "targets";
+
+/// rotation: yaw/pitch
+const ARG_ROTATION: &str = "rotation";
+
+/// single entity
+const ARG_FACING_ENTITY: &str = "facingEntity";
+
+/// position
+const ARG_FACING_LOCATION: &str = "facingLocation";
```

Comment:
I also would like to have the string in the arguments, Its annoying to need add them every time

Reply:
Hi, could you elaborate? I could inline those consts of course, but they have several advantages:
- IDE auto completion
- typos become compiler errors instead of runtime errors
- actual value can be changed very quickly, eg if we migrate from using string keys to enums or such
- you can see all arguments at a glance

Also, for this command, the work is already done. I totally see why you would want to avoid that work for new commands tho, especially if they are less complex.

Reply:
I don't really see the points, I have the feeling you don't understand what i want. We already have like 4 times `ARG_TARGETS` constant. **Which is annoying and can be inconsistent**. So Lets just go to the `ArgumentConsumer` trait and add a const there like `NAME: String`. Then we automatically can use the names from everywhere. I would also make working with translations easier

Reply:
> I have the feeling you don't understand what i want

Yes, that's why I asked you to elaborate :)

I don't think we should translate these at all (and I don't think Minecraft does either), as they are mostly internal identifiers, i.e. their primary use is as keys in a HashMap. That also means that for any given path, each arg needs a unique name. Adding them to ArgumentConsumer won't work, because some commands have paths where several arguments use the same consumer. For example /fill needs a start Pos3d and an end Pos3d and they can't have the same name.

Reply:
1. I also don't think Minecraft does that but Pumpkin wants to have more Features than Vanilla, I think it could be cool to have the args translated, I don't say that this had to be now or in the future but its just an idea, Because there are efforts going on to add Translations to Pumpkin (https://github.com/Snowiiii/Pumpkin/issues/120).
2. Your right, I was thinking of having a default name in the trait and let the user also give the option to choose a custom name, Example:

Will use default:
`parse_arg_position_3d(args)?;`
Will use custom name:
`parse_arg_position_3d("my_cool_custom_position", args)?;`

Because its way more common to use the default than an custom name

Reply:
I actually just wrote a macro to replace those parse functions, so this might be very doable 👍

Reply:
@Snowiiii 

Args can be retrieved in executor using `let distance = get_parsed_arg!(args, ARG_DISTANCE, Arg::F64(v), *v)?;`.
If default name from Consumer should be used: `let Vector2 { x, z } = get_parsed_arg_default!(args, Position2DArgumentConsumer, Arg::Pos2D(vec), *vec)?;`

e.g. `argument_default(&Position2DArgumentConsumer)` can be used in init_command_tree, too.

(All examples are taken from cmd_worldborder.rs)

Reply:
I like `argument_default(&Position2DArgumentConsumer)`. I find the other macros really confusing

Reply:
I added some more documentation to those. All they do is retrieve the parsed argument from the HashMap and if it matches with the given enum variant, return the enum variants content. The same can be done explicitly, but that'd be several more lines of code for each argument.

Reply:
So basically
```
let targets = get_parsed_arg!(args, ARG_TARGETS, Arg::Entities(e), e)?;
```
is the same as 
```
let targets = match args.get(ARG_TARGETS) {
    Some(Arg::Entities(e)) => Ok(e),
    _ => Err(crate::command::dispatcher::InvalidTreeError::InvalidConsumptionError(Some(ARG_TARGETS.into())))
}?;`
```

Reply:
Im still not a big fan of that, The Whole args API just looks pain in the ass and definitely not fun. I want to a have simple to get things.
I could put the args parsing the to arg files again and have it much simpler. Example
`let entites = EntitiesArgumentConsumer::parse(args, ARG_TARGETS)`

Reply:
I previously discarded that option because parse cannot be part of the ArgumentConsumer trait as its return type would be generic and that's not allowed for trait objects. Having parse implemented on the structs directly would work tho, I'll look into that.

Reply:
@Snowiiii I created a FindArg trait and implemented it for all current cosumers and a FindArgDefaultName trait that is automatically implemented for every consumer that implements both DefaultNameArgConsumer and FindArg

Reply:
Better already